internal/newrelic/asserter.go

```diff
@@ -0,0 +1,135 @@
+		return fmt.Errorf("executing nrql query %s, %w", query, err)
+	}
+	if len(a.Results) == 0 {
+		return fmt.Errorf("%w: %s", ErrNoResult, query)
+	}
+	if !validValue(a.Results) {
+		return fmt.Errorf("%w: %s", ErrNotValid, query)
+	}
+	return nil
+}
+
+func validValue(queryResults []nrdb.NRDBResult) bool {
+	firstResult := queryResults[0]
+	for key, val := range firstResult {
+		if key == "timestamp" {
+			continue
+		}
+		if val == nil {
+			return false
+		}
+	}
+	return true
+}
+
+func nrqlQueryExpectedValueTest(nrc *nrClient, query string, expectedResults []spec.TestNRQLExpectedResult) error {
+	a, _ := nrc.client.Query(nrc.accountID, query)
+
+	if len(expectedResults) != len(a.Results) {
+		return fmt.Errorf("%w: %s\n - expected %d got %d", ErrResultNumber, query, len(expectedResults), len(a.Results))
+	}
+	for i, expectedResult := range expectedResults {
+		actualResult := a.Results[i][expectedResult.Key]
+		comparisonErr := compareResults(actualResult, expectedResult)
+		if comparisonErr != nil {
+			return fmt.Errorf("%w: %s\n - for key '%s': %s", ErrNotExpectedResult, query, expectedResult.Key, comparisonErr.Error())
+		}
+	}
+	return nil
+}
+
+func compareResults(actualResult any, expectedResult spec.TestNRQLExpectedResult) error {
+	expectedExactResult := expectedResult.Value
+	if expectedExactResult != nil {
+		// We are checking for an exact value
+		expectedExactResult = preprocessResult(expectedExactResult)
+		actualResult = preprocessResult(actualResult)
+
+		if expectedExactResult == actualResult {
+			return nil
+		}
+		return fmt.Errorf("%w - expected: '%s', got '%s'", ErrAssertionFailure, expectedExactResult, actualResult)
+	}
+
+	// We are checking for a bounded value
+	return checkBounds(actualResult, expectedResult.LowerBoundedValue, expectedResult.UpperBoundedValue)
+}
+
+func preprocessResult(result any) any {
+	switch typedResult := result.(type) {
+	case int:
+		intResult := typedResult
+		// Convert int into floats
+		return float64(intResult)
+	case string:
+		stringResult := typedResult
+		// Convert string nil into nil
+		if strings.EqualFold(stringResult, "nil") {
+			return nil
+		}
+
+		// Convert string booleans into boolean type
+		if strings.EqualFold(stringResult, "false") {
+			return false
+		} else if strings.EqualFold(stringResult, "true") {
+			return true
+		}
+		return stringResult
+	}
+
+	return result
+}
+
+func extractFloat(result any) (float64, error) {
+	result = preprocessResult(result)
+	floatResult, ok := result.(float64)
+	if !ok {
+		return 0, fmt.Errorf("%w: float", ErrTypeAssertion)
+	}
+	return floatResult, nil
+}
+
+func checkBounds(actualResult any, expectedLowerResult *float64, expectedUpperResult *float64) error {
+	actualFloat, err := extractFloat(actualResult)
+	if err != nil {
+		return err
+	}
+
+	// if either expectedLowerResult is nil or expectedLowerResult <= result, AND either expectedUpperResult is nil or result <= expectedUpperResult, return nil, ELSE, error
```

Comment:
Can we remove this comment? It uses almost the same syntax as the actual code below so I don't think it helps to clarify.